### PR TITLE
[Graphbolt]Add threading utils

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add 'threading_utils.h', currently only add one function to compute thread number.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/include/graphbolt/threading_utils.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace graphbolt {
+namespace runtime {
+
+inline size_t compute_num_threads(size_t begin, size_t end, size_t grain_size) {
+#ifdef _OPENMP
+  if (omp_in_parallel() || end - begin <= grain_size || end - begin == 1)
+    return 1;
+
+  return std::min(
+      static_cast<int64_t>(torch::get_num_threads()),
```
Comment (frozenbugs) [inline graphbolt/include/graphbolt/threading_utils.h:20]: I don't see torch related include, this is correct?

File: graphbolt/include/graphbolt/threading_utils.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @file graphbolt/utils.h
+ * @brief Threading utils in graphbolt.
+ */
+#ifndef GRAPHBOLT_THREADING_UTILS_H_
+#define GRAPHBOLT_THREADING_UTILS_H_
+
+#include <omp.h>
+
+namespace graphbolt {
+namespace runtime {
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:12]: as we're adding more and more files in this path, I think we need to split them into further sub-directories, especially the ones lie in different namespaces. for this one, it's in namespace `runtime` while some of other existing files are in namespace `sampling`.

File: graphbolt/include/graphbolt/threading_utils.h (hunk 1 above)
Comment (frozenbugs) [reply graphbolt/include/graphbolt/threading_utils.h:20]: That would be the ideal case. But nowadays most of the c++ code are tested along with Python, so we don't have this requirement.

But all the code should be able to build at least.

Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:20]: the return type of function is  `size_t` while it's cast to `int64_t` here. Mismatch. I don't think they are the same though `size_t` is implementation-defined.

File: graphbolt/include/graphbolt/threading_utils.h
Comment (frozenbugs) [inline graphbolt/include/graphbolt/threading_utils.h:1]: Does this need to be in graphbolt/include? should it be in just src?

File: graphbolt/include/graphbolt/threading_utils.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+inline size_t compute_num_threads(size_t begin, size_t end, size_t grain_size) {
+#ifdef _OPENMP
+  if (omp_in_parallel() || end - begin <= grain_size || end - begin == 1)
+    return 1;
+
+  return std::min(
+      static_cast<int64_t>(torch::get_num_threads()),
+      torch::divup(end - begin, grain_size));
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:21]: same question, could we use `std::ceil` instead?

File: graphbolt/include/graphbolt/threading_utils.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#ifndef GRAPHBOLT_THREADING_UTILS_H_
+#define GRAPHBOLT_THREADING_UTILS_H_
+
+#include <omp.h>
+
+namespace graphbolt {
+namespace runtime {
+
+inline size_t compute_num_threads(size_t begin, size_t end, size_t grain_size) {
+#ifdef _OPENMP
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:15]: will we assume `GraphBolt` utilize OPENMP for parallel in most of time? Does `torch::parallel_for` utilize OpenMP? Will we prefer to use `torch::parallel_for`? If so, do we need to add `#ifdef` here?

File: graphbolt/include/graphbolt/threading_utils.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ */
+#ifndef GRAPHBOLT_THREADING_UTILS_H_
+#define GRAPHBOLT_THREADING_UTILS_H_
+
+#include <omp.h>
+
+namespace graphbolt {
+namespace runtime {
+
+inline size_t compute_num_threads(size_t begin, size_t end, size_t grain_size) {
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:14]: the naming style is what we're following now?

File: graphbolt/include/graphbolt/threading_utils.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#ifdef _OPENMP
+  if (omp_in_parallel() || end - begin <= grain_size || end - begin == 1)
+    return 1;
+
+  return std::min(
+      static_cast<int64_t>(torch::get_num_threads()),
+      torch::divup(end - begin, grain_size));
+#else
+  return 1;
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/threading_utils.h:23]: curious only: I am not aware of the scenario of calling this function(at least from this PR), so I don't know if it's correct to return `1` for `#else`.

**Pull request closed without merging**